feat(cli): warn of non-existent stacks in `cdk destroy`

go-to-k opened the pull request:
Fixes https://github.com/aws/aws-cdk/issues/32836, https://github.com/aws/aws-cdk/issues/32545, https://github.com/aws/aws-cdk/issues/27179, https://github.com/aws/aws-cdk/issues/22240

## Reason for this change

This PR implements the feature that warns users when non-existent stacks are specified in `cdk destroy`.

* It does not display the message `Are you sure you want to delete:` if there is no matching stack.
* Even if the stack does not exist, `cdk destroy` will not fail, it will just print a warning.

For examples (that have `Stacka`, `StackA`, `StackX`):

<img width="355" height="56" alt="destroy2" src="https://github.com/user-attachments/assets/0ac39df5-7ec7-46a4-8572-f61bb7f26ec6" />
<br>
<img width="414" height="85" alt="destroy3" src="https://github.com/user-attachments/assets/9c897485-71a9-4c93-9d00-4ed5a5f8df15" />
<br>
<img width="315" height="56" alt="destroy4" src="https://github.com/user-attachments/assets/d208e590-f3c8-4228-87df-0797202a6224" />

## Difference from previous PR

The [previous PR](https://github.com/aws/aws-cdk/pull/32636) was reverted in https://github.com/aws/aws-cdk/pull/32839 due to a regression with only nested stage stacks. So this version addresses that regression with comprehensive tests.

## Description of changes

The original implementation added warnings for non-existent stacks in `cdk destroy`, but it failed when applications had only nested stage stacks (no top-level stacks). This happened because the code used `allTopLevel: true`, which only searched for stacks directly under the App, ignoring stacks within nested Stages.

Fixed the regression by changing `suggestStacks` method to use `DefaultSelection.AllStacks` instead of `allTopLevel: true`, ensuring the warning feature works for all stack configurations (top-level only, nested only, or both). Added regression tests to verify the fix for the nested stage scenario.

```diff
  private async suggestStacks(props: {
    selector: StackSelector;
    stacks: StackCollection;
    exclusively?: boolean;
  }) {
    const assembly = await this.assembly();
    const selectorWithoutPatterns: StackSelector = {
-      ...props.selector,
-      allTopLevel: true,
      patterns: [],
    };
    const stacksWithoutPatterns = await assembly.selectStacks(selectorWithoutPatterns, {
      extend: props.exclusively ? ExtendedStackSelection.None : ExtendedStackSelection.Downstream,
-      defaultBehavior: DefaultSelection.OnlySingle,
+      defaultBehavior: DefaultSelection.AllStacks,
    });
```

## Additional Information

When running `cdk destroy --all` or `cdk deploy --all` against a configuration with **no top-level stacks (nested stages only)**, the following error occurs.

However, this behavior existed **before this PR** and is unrelated to the changes made here. Since it is outside the scope of this PR, no fix has been implemented for this behavior.

FYI: I have submitted an [issue](https://github.com/aws/aws-cdk-cli/issues/1003) and a [PR](https://github.com/aws/aws-cdk-cli/pull/985) about this behavior.

```ts
const app = new cdk.App();

new MyStage(app, 'MyStage'); // This has CdkSampleStack
```

```
> cdk deploy
Since this app includes more than a single stack, specify which stacks to use (wildcards are supported) or specify `--all`
Stacks: MyStage/CdkSampleStack

> cdk deploy --all
...
No stack found in the main cloud assembly. Use "list" to print manifest

> cdk destroy --all
...
No stack found in the main cloud assembly. Use "list" to print manifest
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

otaviomacedo committed:
refactor: filter stacks before getting their templates (#960)

The `refactor` command only operates on the stacks that are relevant for
the target CDK application. However, it gets all the templates of the
deployed stacks before filtering them, which is wasteful.

Invert the order, so that the filtering happens before getting the
templates.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

go-to-k committed:
destroy

go-to-k committed:
color codes

go-to-k committed:
fix the previous bug

go-to-k committed:
add tests

comment

go-to-k committed:
skip --all

go-to-k committed:
add tests

go-to-k committed:
add tests

go-to-k committed:
integ

go-to-k committed:
refactor tests

test titles

test tiltes

modify tests

allTopLevel in tests

modify tests

rm test

add test

title

go-to-k committed:
modify

go-to-k committed:
refactor

go-to-k committed:
modify

go-to-k committed:
tweak

go-to-k committed:
title

go-to-k committed:
refactor tests to separate methods

go-to-k committed:
rm unused

go-to-k committed:
integ

tweak

go-to-k committed:
fix integ

go-to-k committed:
test

go-to-k committed:
style

go-to-k committed:
integ namte

go-to-k committed:
add unit test to coverage

go-to-k committed:
require exclusively

go-to-k committed:
change integ

integ

integ

integ

go-to-k committed:
refactor tests

go-to-k committed:
Merge branch 'main' of https://github.com/go-to-k/aws-cdk-cli into destroy

go-to-k committed:
refactor: route cdk destroy through toolkit-lib

toolkit-lib: add `force` to DestroyOptions; add W7010/W7011 warnings; move the
stack-name suggestion + no-match warning + force-skip into `_destroy`; expose
`destroyForAction` as protected so the CLI can keep `deploy` message attribution.

aws-cdk: thin destroy delegation with selector -> StackSelector mapping; add
`InternalToolkit.destroyFromDeploy`; remove `fromDeploy`/legacy loop; preserve
graceful abort on a declined confirmation.

go-to-k committed:
test: migrate destroy tests to toolkit-lib

Move the destroy logic tests to toolkit-lib (including stage/nested-stage
regression coverage and the force / suggestion / no-match cases) with new stage
fixtures; reduce the CLI tests to delegation/mapping checks; rework the destroy
command snapshot test to spy on Deployments.prototype.

go-to-k committed:
chore: regenerate message registry and destroy IoHost snapshots

Generated output of routing destroy through toolkit-lib: message-registry.md is
regenerated from messages.ts, and the destroy command's __io_snapshots__ reflect
the messages now emitted by toolkit-lib. No hand edits.

go-to-k committed:
chore(toolkit-lib): fix picomatch import order (eslint import/order)

go-to-k committed:
fix(cli): make destroy IoHost snapshots color-independent

The destroy command snapshot test was non-deterministic across color modes:
toolkit-lib wraps a leading newline inside a chalk color escape (`\n ✅ ...`),
so `withTrimmedWhitespace`'s `.trim()` cannot remove it when colors are on (the
string starts with the ANSI escape, not whitespace). The recorder stripped ANSI
only after receiving the already-(un)trimmed message, so the leading newline
survived with colors on and the snapshot differed between a TTY and CI.

Trim in the recorder's `normalize()` after stripping ANSI so the recorded stream
is the same regardless of whether colors were enabled, and regenerate the
affected snapshots.

go-to-k committed:
fix(toolkit-lib): keep destroy confirmation stack names blue

The destroy confirmation prompt colored the stack names red. Use blue to match
the color the CLI used before destroy was routed through toolkit-lib.

go-to-k committed:
Merge branch 'main' into destroy

Align cdk destroy with the confirmation-prompt contract from #1667:
declining now throws AbortError('DestroyAborted', 'Deletion cancelled')
and exits non-zero instead of emitting CDK_TOOLKIT_E7010 and exiting 0.

github-actions[bot] committed:
chore: self mutation

Signed-off-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

go-to-k commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -1593,18 +1604,28 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+    if (stacks.stackCount === 0) {
+      await ioHelper.notify(IO.CDK_TOOLKIT_W7011.msg(
+        `No stacks match the name(s): ${chalk.red((selectStacks.patterns ?? []).join(', '))}`,
+      ));
```
Note on the exit code: no-match (W7011) exits 0, while a declined confirmation exits 1 (#1667). This is deliberate and keeps the existing behavior. The old `selectStacksForDestroy` had a `// No validation` comment and returned an empty selection, so `cdk destroy <nonexistent>` already exited 0 (deploy throws `NoStacksMatched`, destroy never did). We keep that contract and add a warning in place of the old empty prompt.

<img width="254" height="121" alt="exit" src="https://github.com/user-attachments/assets/a8dcb5a6-3471-4dda-8804-18f269e12494" />

mrgrain replied:
yes that's great!

go-to-k committed:
Merge branch 'main' of https://github.com/go-to-k/aws-cdk-cli into destroy

# Conflicts:
#	packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts
#	packages/aws-cdk/lib/cli/cdk-toolkit.ts
#	packages/aws-cdk/test/_helpers/io-recorder.ts
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/destroy_failure_emits_a_failure_message_and_rethrows_when_destroyStack_fails.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/force_false_confirmation_prompt_aborts_with_an_AbortError_and_destroys_nothing_when_the_user_declines.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/force_false_confirmation_prompt_asks_for_confirmation_and_proceeds_when_the_user_confirms.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/force_false_confirmation_prompt_rethrows_an_unexpected_error_from_the_confirmation_prompt.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/force_true_no_confirmation_prompt_destroys_a_single_nested_stack_fromDeploy_makes_it_say_deployed.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/force_true_no_confirmation_prompt_destroys_all_top-level_stacks_with_concurrency.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/force_true_no_confirmation_prompt_forwards_the_roleArn_to_destroyStack.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/destroy/force_true_no_confirmation_prompt_respects_dependency_order_with_concurrency.ndjson
#	packages/aws-cdk/test/commands/__io_snapshots__/list/cdk_list_prints_machine-readable_JSON_without_the_synthesis-time_line.ndjson
#	packages/aws-cdk/test/commands/destroy.test.ts

go-to-k committed:
feat(toolkit-lib): destroy warns and suggests when stack names do not match

Re-applies the destroy warning feature on top of main's toolkit-lib destroy
migration (#1686):
- W7011: emit a warning and exit without prompting when no stacks match
- W7010: warn about each non-existent name (suggesting a close match when one
  exists, including stacks nested in a stage) while still destroying matches
- adds suggestStacks() to Toolkit and the supporting message codes/registry
- restores toolkit-lib unit tests, stage fixtures, and the destroy integ tests

The toolkit-lib migration plumbing from this branch is dropped in favor of
main's implementation; only the warning behavior is preserved.

go-to-k committed:
fix(toolkit-lib): keep destroy confirmation stack names blue

The destroy confirmation prompt colored the stack names red (inherited from
main's _destroy). Use blue to match the color the CLI used before destroy was
routed through toolkit-lib, and assert it so the merge does not silently
re-introduce red.

Ref: https://github.com/aws/aws-cdk-cli/pull/984#issuecomment-4787129749

go-to-k committed:
test(toolkit-lib): drop redundant comment in destroy confirmation color test

go-to-k committed:
Merge branch 'main' into destroy

go-to-k committed:
Merge branch 'main' into destroy

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -1692,6 +1702,39 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+   * when one exists (e.g. only the casing differs). Stacks nested inside a stage
+   * are considered, so a staged stack can be suggested too.
+   */
+  private async suggestStacks(
+    ioHelper: IoHelper,
+    assembly: StackAssembly,
+    selector: StackSelector,
+    selected: StackCollection,
+  ): Promise<void> {
+    const patterns = selector.patterns ?? [];
+    if (patterns.length === 0) {
+      return;
+    }
+
+    const allStacks = await assembly.selectStacksV2(ALL_STACKS);
+
+    for (const pattern of patterns) {
+      const matched = selected.stackArtifacts.some((stack) => picomatch.isMatch(stack.hierarchicalId, pattern));
+      if (matched) {
+        continue;
+      }
+
+      const closeMatches = allStacks.stackArtifacts
+        .filter((stack) => picomatch.isMatch(stack.hierarchicalId.toLowerCase(), pattern.toLowerCase()))
+        .map((stack) => stack.hierarchicalId);
```
This function to match candidates (basically to this line) should live on `StackAssembly` class. It should not log any output but return a `string[]`.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -1692,6 +1702,39 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+    const allStacks = await assembly.selectStacksV2(ALL_STACKS);
+
+    for (const pattern of patterns) {
+      const matched = selected.stackArtifacts.some((stack) => picomatch.isMatch(stack.hierarchicalId, pattern));
+      if (matched) {
+        continue;
+      }
```
Can you explain what purpose this is serving? I just don't understand it. We are excluding any of the stacks that already match the pattern themselves? But why?

go-to-k replied:
This handles the case where several names are given and only some exist. E.g. `cdk destroy StackA Typo`: `StackA` is destroyed, `Typo` matches nothing. The `matched` check skips `StackA` and warns only about `Typo` ("Did you mean …?").

With "only warn when 0 stacks matched", a typo mixed in with valid names would slip through unnoticed, that's what I wanted to catch. That said, if you'd prefer a single message (only when nothing matched at all), I'm happy to fold it into one payload-carrying message. Which do you prefer?

mrgrain replied:
Okay, thank you. I understand now what this is trying to achieve. Let me think about this for a bit. Some of my comments are now obviously contingent on this decision here.

go-to-k replied:
Holding off until the direction above is settled, then I'll adjust accordingly.

go-to-k replied:
@mrgrain How has it been since then?

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -1620,12 +1621,21 @@ export class Toolkit extends CloudAssemblySourceBuilder {
     const ioHelper = asIoHelper(this.ioHost, action);
     const stacks = await assembly.selectStacksV2(selectStacks);
 
+    await this.suggestStacks(ioHelper, assembly, selectStacks, stacks);
```
I think we should only call this if the original selector returned 0 results.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -1620,12 +1621,21 @@ export class Toolkit extends CloudAssemblySourceBuilder {
     };
 
+    if (stacks.stackCount === 0) {
+      await ioHelper.notify(IO.CDK_TOOLKIT_W7011.msg(
```
I think both messages can be folded into a single message that informs the user that no matching stacks were found. And if we have identified possible close matches ("candidates"), then we suggest them ("Did you mean...").

The Toolkit message should include a payload of the unmatched pattern and the identified matches as `string[] | undefined`.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/io/private/messages.ts:
```diff
@@ -398,6 +398,15 @@ export const IO = {
     interface: 'StackDestroy',
   }),
 
+  CDK_TOOLKIT_W7010: make.warn({
+    code: 'CDK_TOOLKIT_W7010',
+    description: 'A provided stack name does not match any stack',
+  }),
+  CDK_TOOLKIT_W7011: make.warn({
+    code: 'CDK_TOOLKIT_W7011',
+    description: 'No stacks match the provided names, nothing to destroy',
+  }),
```
Like said elsewhere, I think a single message will be enough but it should carry a payload of the unmatched selector and idenfified matches.

mrgrain reviewed (changes requested):
Thanks for your continued work! FYI I want to let the changes to `destroy` bake a little bit longer to see if there's any issue from that.

But this PR is getting close to be ready now. Made some final comments on code organisation and messaging. We will also expect new tests in `packages/aws-cdk/test/commands/destroy.test.ts` and associated snapshots to record the exact output.

go-to-k committed:
Merge branch 'main' into destroy

go-to-k committed:
Merge branch 'main' into destroy

go-to-k commented:
@mrgrain How has it been since then?

go-to-k committed:
Merge branch 'main' into destroy

go-to-k committed:
Merge branch 'main' into destroy

mrgrain reviewed (changes requested):
Right. I'd like to encapsulate the suggestion mechanism inside the `StackAssembly` class.

- We introduce a new `selectStacksV3` method that replaces the current `selectStacksV2`.
- This will get a new options bag as second argument. For now, we have one option `suggestPatternMatches: boolean // default: false`
- The return will change to an object: `stacks: StackCollection`
-When `suggestPatternMatches === true` the return value also includes `suggestions: Recod<pattern: string, string[]>` if the stack selector is a pattern selector and no stack was found for a given pattern.
- The existing `selectStacksV2` becomes a thin wrapper around `selectStacksV3` to maintain the current API. Let's not migrate existing usage just yet, we can do a follow up.
- In `_destroy` we start using `selectStacksV3`. We then use the information to craft messages. 

Start with this, then we can do another round. Sorry this took so long.

go-to-k committed:
refactor(toolkit-lib): move destroy match suggestions into StackAssembly.selectStacksV3

Encapsulate the pattern-match suggestion mechanism inside StackAssembly, per
review feedback on #984:
- add selectStacksV3(selector, { suggestPatternMatches }) returning
  { stacks, suggestions? } where suggestions is a Record<pattern, string[]> of
  close matches for each pattern that selected no stack
- selectStacksV2 becomes a thin wrapper over selectStacksV3 (existing callers
  unchanged; migrating them is a follow-up)
- _destroy consumes selectStacksV3 and crafts the W7010/W7011 messages from the
  returned info; the standalone Toolkit.suggestStacks helper is removed

Message consolidation into a single payload-carrying message is a follow-up.

go-to-k commented:
@mrgrain 

okay, changed it.

mrgrain committed:
small refactor

mrgrain committed:
Merge branch 'main' into destroy

mrgrain commented:
:shipit:
